di/_utils/task.py

```diff
@@ -134,119 +140,266 @@ def generate_execute_fn(
     def __repr__(self) -> str:
         return f"{self.__class__.__name__}(scope={self.scope}, call={self.call})"
 
+    # This is the general execution path that happens the first time this Task/dependency is executed
+    # We call the dependency then introspect the result to determine what sort of dependency it is
+    # Once it knows that, this function replaces itself with a specialized variant
 
-class AsyncTask(Task, ExecutorAsyncTask):
-    __slots__ = ("is_generator",)
-
-    def __init__(
-        self,
-        scope: Scope,
-        call: DependencyProvider,
-        use_cache: bool,
-        dependant: DependantBase[Any],
-        task_id: int,
-        positional_parameters: Iterable[Union[AsyncTask, SyncTask]],
-        keyword_parameters: Iterable[Tuple[str, Union[AsyncTask, SyncTask]]],
-    ) -> None:
-        super().__init__(
-            scope,
-            call,
-            use_cache,
-            dependant,
-            task_id,
-            positional_parameters,
-            keyword_parameters,
-        )
-        self.is_generator = is_async_gen_callable(self.call)
-        if self.is_generator:
-            self.call = asynccontextmanager(self.call)  # type: ignore[arg-type]
-
-    async def compute(  # type: ignore[override]  # we do it this way to avoid exposing implementation details to users
+    def unspecialized(
```

Comment:
@graingert I know it's hard to see what's going on in this diff. But in pseudocode the logic looks something like this:

```python
async_stack = AsyncExitStack()
sync_stack = ExitStack()

res = user_function_or_class()

if inspect.isawaitable(res):
    # must be a coroutine function
    return await res
if hasattr(res, "__aenter__"):
    # must be an async context manager
    return await async_stack.enter_async_context(res)
if hasattr(res, "__enter__"):
    # must be a sync context manager
    return sync_stack.enter_context(res)
```

Where `user_function_or_class` can be any of:

```python
def func(): ...

async def func(): ...

@contextmanager
def func():  yield

@asynccontextmanager
async def func(): yield


class CM:
    def __enter__(self): ...
    def __exit__(...): ...

class AsyncCM:
    async def __aenter__(self): ...
    async def __aexit__(self): ...
```

Does this sound safe/correct to you?